src/bigistore.c

```diff
@@ -1058,7 +1056,52 @@ Datum bigistore_slice(PG_FUNCTION_ARGS)
         PG_RETURN_POINTER(result);
     }
     else
-        PG_RETURN_NULL();
+        PG_RETURN_EMPTY_ISTORE();
+}
+
+PG_FUNCTION_INFO_V1(bigistore_slice_min_max);
+Datum bigistore_slice_min_max(PG_FUNCTION_ARGS)
+{
+    int32      min = PG_GETARG_INT32(1);
+    int32      max = PG_GETARG_INT32(2);
+    BigIStore *is  = PG_GETARG_BIGIS_COPY(0);
+
+    int min_idx = 0;
+    int i       = 0;
+    int len     = is->len;
+
+    BigIStorePair *pairs;
+    pairs = FIRST_PAIR(is, BigIStorePair);
+
+    if (min > max)
+        ereport(ERROR, (errcode(ERRCODE_SYNTAX_ERROR), errmsg("min must be less or equal max")));
+
+    if (is->len <= 0 || pairs[is->len - 1].key < min || pairs[0].key > max)
+        PG_RETURN_EMPTY_ISTORE();
+
+    if (pairs[0].key >= min && pairs[is->len - 1].key <= max)
+        PG_RETURN_POINTER(is);
+
+    is->buflen = 0;
+    is->len    = 0;
+    // skip pairs lower than min
+    while (pairs[i].key < min && ++i)
+        ;
+    min_idx = i;
+
+    for (; pairs[i].key <= max && i < len; i++)
+    {
+        ++is->len;
+        is->buflen += bigis_pair_buf_len(pairs + i);
```

Comment:
I just realized the index is referenced at line 1095, so by incrementing it beforehand in the loop we will skip elements examined in the loop conditions and take the next one, that doesn't necessary satisfy those (or even exists). That is, if i is 1 after the while loop, the for loop will check the pairs[1], but the buflen will are incremented by the pairs[2].

We should write it as
```
while (pairs[i].key <= max && i < len)
{
    ++is->len
    is->buflen += bigis_pairs_buf_len(pairs + i++)
}
```

Reply:
NM, As @zilder  pointed out this is not how the for loop works :-)